nupic/research/frameworks/pytorch/datasets/permuted_mnist.py

```diff
@@ -0,0 +1,146 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2021, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import os
+
+import torch
+from torchvision import transforms
+from torchvision.datasets import MNIST
+
+from nupic.research.frameworks.dendrites.routing import generate_context_vectors
+
+
+class PermutedMNIST(MNIST):
+    """
+    The permutedMNIST dataset contains MNIST images where the same random permutation
+    of pixel values is applied to each image. More specifically, the dataset can be
+    broken down into 'tasks', where each such task is the set of all MNIST images, but
+    with a unique pixel-wise permutation applied to all images. `num_tasks` gives the
+    number of 10-way classification tasks (each corresponding to a unique pixel-wise
+    permutation) that a continual learner will try to learn in sequence.
+    """
+
+    # Permutations should be the same whether this class is instantiated for training
+    # or testing
+    permutations = None
+
+    def __init__(self, num_tasks, train, root=".", target_transform=None,
+                 download=False):
+
+        data_transform = transforms.Compose([
+            transforms.ToTensor(),
+            transforms.Normalize((0.13062755,), (0.30810780,)),
+        ])
+        super().__init__(root=root, train=train, transform=data_transform,
+                         target_transform=target_transform, download=download)
+
+        self.num_tasks = num_tasks
+
+        # Generate num_tasks random permutations - the first one is the identity
+        # permutation (i.e., regular MNIST), represented below as `None`
+        if PermutedMNIST.permutations is None:
+            PermutedMNIST.permutations = [
```

Comment:
`permutations` is clearly an instance variable, since it is modified in the init method of a class (the init initializes an instance). Why is it defined as a class variable?